compiler-plugin/src/main/kotlin/arrow/meta/higherkind/HigherKindPlugin.kt

```diff
@@ -24,16 +23,19 @@ val MetaComponentRegistrar.higherKindedTypes: List<ExtensionPhase>
             "typealias ${name}PartialOf<${c.partialTypeParameters}> = arrow.Kind${c.partialKindAritySuffix}<For$name, ${c.partialTypeParameters}>"
           else null,
           /** Class redefinition with kinded super type **/
+          // Warning needs to be added, because of cases where one ADT Branch is isomorphic in respect to the ADT, e.g.:
+          // Kind<ForExpr, Int> == ForSum == Sum<C> == Expr<Int, C>
           """
-              |$modality $visibility $kind $name<$typeParameters>($valueParameters) : ${name}Of<${typeParameters.invariant}> {
+              |@Suppress("INCONSISTENT_TYPE_PARAMETER_VALUES")
+              |$visibility $modality $kind $name<$typeParameters>($valueParameters) : ${if ((String::isNotEmpty)(supertypes.identifier)) ({ it: String -> "$it, " })(supertypes.identifier) else supertypes.identifier}${name}Of<${typeParameters.invariant}> {
               |  $body
               |}
-              |"""
+              """.trimMargin()
         )
       }
     )
 
-private val Name.invariant: String
+val Name.invariant: String
```

Comment:
All these functions are unrelated to autofold and should be moved to the KtUtils.